File: week01_C/pset/credit/credit.c
#include <cs50.h>
#include <math.h>
#include <stdio.h>

int checksum(long n);
int count(long n);

int main(void)
{
    long n = get_long("Number: ");
    long x = pow(10, count(n) - 1);
    long y = pow(10, count(n) - 2);

    if (checksum(n) == 0)
    {
        if (count(n) == 15 && (n / y == 34 || n / y == 37))
        {
            printf("AMEX\n");
        }

        else if ((count(n) == 13 || count(n) == 16) && n / x == 4)
        {
            printf("VISA\n");
        }
        else if (count(n) == 16 && (n / y == 51 || n / y == 52 || n / y == 53 || n / y == 54 || n / y == 55))
        {
            printf("MASTERCARD\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

// Count digits in number
int count(long n)
{
    int c = 0;
    while (n > 0)
    {
        c++;
        n = n / 10;
    }
    return c;
}

// Calculate checksum
int checksum(long n)
{
    int r = 0;
    int t = 0;
    for (int i = count(n); i > 0; i = i - 2)
    {
        r += n % 10;
        n /= 10;
        int s = (n % 10) * 2;
        t += s / 10 + s % 10;
        n /= 10;
    }
    int u = (r + t) % 10;
    return u;
}

File: week06_Python/pset/sentimental-credit/credit.py
# TODO
import re


def main():
    n = input("Number: ")
    pattern_AMEX = r'^(34|37)\d{13}$'
    pattern_VISA = r'^4\d{12,15}$'
    pattern_MASTER = r'^5[1-5]\d{14}$'

    if checksum(n) == 0:
        if re.match(pattern_AMEX, n):
            print('AMEX')
        elif re.match(pattern_VISA, n):
            print('VISA')
        elif re.match(pattern_MASTER, n):
            print('MASTERCARD')
        else:
            print('INVALID')
    else:
        print('INVALID')


def checksum(n):
    ni = int(n)
    r = s = t = 0
    for i in range(len(n), 0, -2):
        r += ni % 10
        ni //= 10
        s = (ni % 10) * 2
        t += int(s / 10) + int(s % 10)
        ni //= 10
    u = (r + t) % 10
    return u


main()
